org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/DefaultModelProvider.java

```diff
@@ -40,40 +47,118 @@
+            @SuppressWarnings("unchecked")
+            T result = (T) this.cache.getIfPresent(cacheKey);
+            return result;
+        }
+
+        // if model must be reloaded, we can invalidate the cache entry and then proceed as for
+        // FetchStrategy.LOAD_IF_NOT_CACHED
+        if (FetchStrategy.FORCE_RELOAD == fetchStrategy) {
+            this.cache.invalidate(cacheKey);
+        }
+
+        // load the values from the cache iff not already cached
+        final AtomicBoolean modelLoaded = new AtomicBoolean(false);
+        T value = getFromCache(cacheKey, new Callable<T>() {
+
+            @Override
+            public T call() {
+                T model = operation.get();
+                modelLoaded.set(true);
+                return model;
+            }
+        });
+
+        return value;
+    }
+
+    private <U> U getFromCache(Class<?> cacheKey, Callable<U> cacheValueLoader) {
+        try {
+            @SuppressWarnings("unchecked")
+            U result = (U) this.cache.get(cacheKey, cacheValueLoader);
+            return result;
+        } catch (Exception e) {
+            throw new GradlePluginsRuntimeException(e);
+        }
     }
 
-    private final TransientRequestAttributes getTransientRequestAttributes(CancellationToken token, IProgressMonitor monitor) {
+    private boolean supportsCompositeBuilds(CancellationToken token, IProgressMonitor monitor) {
+        BuildEnvironment buildEnvironment = fetchModel(BuildEnvironment.class, FetchStrategy.FORCE_RELOAD, token, monitor);
+        GradleVersion gradleVersion = GradleVersion.version(buildEnvironment.getGradle().getGradleVersion());
+        return gradleVersion.getBaseVersion().compareTo(GradleVersion.version("3.3")) >= 0;
+    }
+
+    private static TransientRequestAttributes getTransientRequestAttributes(CancellationToken token, IProgressMonitor monitor) {
         ProcessStreams streams = CorePlugin.processStreamsProvider().getBackgroundJobProcessStreams();
-        List<ProgressListener> progressListeners = ImmutableList.<ProgressListener> of(DelegatingProgressListener.withoutDuplicateLifecycleEvents(monitor));
-        ImmutableList<org.gradle.tooling.events.ProgressListener> noEventListeners = ImmutableList.<org.gradle.tooling.events.ProgressListener> of();
+        List<ProgressListener> progressListeners = ImmutableList.<ProgressListener>of(DelegatingProgressListener.withoutDuplicateLifecycleEvents(monitor));
+        ImmutableList<org.gradle.tooling.events.ProgressListener> noEventListeners = ImmutableList.<org.gradle.tooling.events.ProgressListener>of();
         if (token == null) {
             token = GradleConnector.newCancellationTokenSource().token();
         }
         return new TransientRequestAttributes(false, streams.getOutput(), streams.getError(), streams.getInput(), progressListeners, noEventListeners, token);
     }
 
-}
+    public static ModelProvider forAttributes(FixedRequestAttributes attributes) {
```

Comment:
There should be no static cache here, that's a memory leak. The ModelProvider should be a field on the GradleBuild.